jerry-core/ecma/builtin-objects/ecma-builtin-global.c

```diff
@@ -361,32 +297,25 @@ ecma_builtin_global_object_decode_uri_helper (lit_utf8_byte_t *input_start_p, /*
       if (lit_is_code_point_utf16_high_surrogate (cp)
           || lit_is_code_point_utf16_low_surrogate (cp))
       {
-        ret_value = ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid UTF8 codepoint."));
-        break;
+        ecma_stringbuilder_destroy (&builder);
+        return ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid UTF8 codepoint."));
       }
 
-      output_char_p += lit_code_point_to_cesu8 (cp, output_char_p);
+      static lit_utf8_byte_t result_chars[6];
+      lit_utf8_size_t cp_size = lit_code_point_to_cesu8 (cp, result_chars);
+      ecma_stringbuilder_append_raw (&builder, result_chars, cp_size);
     }
   }
 
-  if (ecma_is_value_empty (ret_value))
-  {
-    JERRY_ASSERT (output_start_p + output_size == output_char_p);
+  lit_utf8_byte_t *builder_buffer_p = ecma_stringbuilder_get_data (&builder);
+  lit_utf8_size_t builder_size = ecma_stringbuilder_get_size (&builder);
 
-    if (lit_is_valid_cesu8_string (output_start_p, output_size))
-    {
-      ecma_string_t *output_string_p = ecma_new_ecma_string_from_utf8 (output_start_p, output_size);
-      ret_value = ecma_make_string_value (output_string_p);
-    }
-    else
-    {
-      ret_value = ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid CESU8 string."));
-    }
+  if (!lit_is_valid_cesu8_string (builder_buffer_p, builder_size))
+  {
+    return ecma_raise_uri_error (ECMA_ERR_MSG ("Invalid CESU8 string."));
```

Comment:
This check seems redundant to me. If the individual code points were valid, the resulting cesu8 encoded string should also be valid.  Do we have/can we make a test that triggers this error?